### PR TITLE
Add status indicator on main home screen for each repo

### LINKED ISSUE #15620: Feature request: Add CI/CD indicator on main home screen for each repo
Right now when viewing a repo, The main line shows the Author, commit hash and commit msg.  If CI/CD is configured (Drone in this case) then a green checkmark (hopefully) is also shown.
image

This is great - shows at a glance that the CI/CD pipeline(s) passed and all is well.

It would be really nice to ALSO see that green checkmark (or whatever status) on the main home screen next to each repo as well.  Would give an overall-status-at-a-glance
image

If it was live, as it is in the repo view, even better.  You glance at the home screen, see a problem, click, click and you're right in the CI/CD to see where the problem is.

Perhaps put it just to the left of the repo name, left of the small icon already there.

### PR DESCRIPTION
**Pull request by yardenshoham**
It will show the calculated commit status state of the latest commit on the default branch for each repository in the dashboard repo list

- Closes #15620

# Before
image

# After
image

### CODE REVIEW

**Comment by yardenshoham** [comment]
I'm aware the code is not good but it works. I'm open to suggestions on how to make it better. We should get this PR to the minimally acceptable state and iteratively make this feature better with additional PRs.

File: modules/structs/repo.go
Comment (delvh) [inline modules/structs/repo.go]:
```suggestion
	LatestCommitStatusState CommitStatusState `json:"latest_commit_status_state"`
```
Sounds at least less confusing to me.

File: web_src/js/components/DashboardRepoList.vue
Comment (delvh) [inline web_src/js/components/DashboardRepoList.vue]: Should we extract a subtemplate perhaps?

Ah wait…
One thing is Vue, the other Go templates.
Hmm…
A custom component instead?

File: routers/web/repo/repo.go
Comment (delvh) [inline routers/web/repo/repo.go]: Functionality-wise I'm all for this PR.
However, I can imagine some huge performance hits for larger repos/ instances…

File: templates/swagger/v1_json.tmpl
Comment (delvh) [inline templates/swagger/v1_json.tmpl]: Hmm… The repository is returned quite a few times.
In all these instances, the latest commit state is empty.
I don't think I can allow this bug to be merged.

**Review by delvh — DISMISSED** [review]

File: templates/swagger/v1_json.tmpl
Comment (yardenshoham) [reply templates/swagger/v1_json.tmpl]: What should I do if I only need this for the dashboard repo list?

Comment (delvh) [reply templates/swagger/v1_json.tmpl]: That is a good question where I'm not sure of the answer myself…
In the worst case a wrapper struct?

File: routers/web/repo/repo.go
Comment (yardenshoham) [reply routers/web/repo/repo.go]: I didn't find a function that does this. In fact, I got lost reading so many functions from modules/models/services of git/repository/branch...

I'm doing weird stuff here like iterating over all branches when I only need the default branch. I hope someone with more knowledge in the codebase help make this better

File: web_src/js/components/DashboardRepoList.vue
Comment (yardenshoham) [reply web_src/js/components/DashboardRepoList.vue]: I could do something very similar to `web_src/js/components/ActionRunStatus.vue` but it might be overkill... The duplication will remain anyway and we won't use it in any other vue component

File: templates/swagger/v1_json.tmpl
Comment (yardenshoham) [reply templates/swagger/v1_json.tmpl]: Hmmm... That would make this a breaking PR for that search endpoint if I create a new struct with 2 fields
```go
type DashboardRepository struct {
	Repository *Repository `json:"repository"`
	LatestCommitStatusState CommitStatusState `json:"latest_commit_status_state"`
}
```

Is there a way to create a struct by extending another? Something like
```go
type DashboardRepository & Repository struct {
	LatestCommitStatusState CommitStatusState `json:"latest_commit_status_state"`
}
```

File: modules/structs/repo.go
Comment (wxiaoguang) [inline modules/structs/repo.go]: It changes the API  (swagger/v1_json.tmpl) ?

If it is only used by web, the API should be kept stable (as before)

Comment (yardenshoham) [reply modules/structs/repo.go]: The DashboardRepoList uses the API to search repositories

Comment (wxiaoguang) [reply modules/structs/repo.go]: The JS can't call API directly, they should have been decoupled ......

Comment (yardenshoham) [reply modules/structs/repo.go]: I'm mistaken, it's `/repo/search?sort=updated&order=desc&uid=1&team_id=undefined&q=&page=1&limit=15&mode=&archived=false`
https://github.com/go-gitea/gitea/blob/23ae939ef3ef03848038372de21490725855b5f9/routers/web/web.go#L794

File: templates/swagger/v1_json.tmpl
Comment (yardenshoham) [reply templates/swagger/v1_json.tmpl]: Not modifying the API anymore

File: modules/structs/repo.go
Comment (yardenshoham) [reply modules/structs/repo.go]: Not modifying the API anymore

File: web_src/js/components/DashboardRepoList.vue
Comment (silverwind) [reply web_src/js/components/DashboardRepoList.vue]: Maybe an object lookup? Should run before or during render.

```js
const commitStatusInfo = {
  pending: ['octicon-dot-fill', 'grey'],
}

const [icon, color] = commitStatusIcons[state];
```

File: routers/web/repo/repo.go
Comment (lunny) [inline routers/web/repo/repo.go]: We need a batch SQL to do that, otherwise it will slow the dashboard.

Comment (yardenshoham) [reply routers/web/repo/repo.go]: How?

File: web_src/js/components/DashboardRepoList.vue
Comment (yardenshoham) [reply web_src/js/components/DashboardRepoList.vue]: Added lookup

Comment (silverwind) [reply web_src/js/components/DashboardRepoList.vue]: Thanks!

File: routers/web/repo/repo.go
Comment (silverwind) [reply routers/web/repo/repo.go]: To me this sounds like a prime use cases for a redis cache. When a commit status changes on main branch or a repo is renamed or deleted, update a redis hash or key with the repo name and the commit state for quick lookup on this list, e.g.

```
HSET commit-statuses go-gitea/gitea 2
HGET commit-statuses go-gitea/gitea

SET commit-statuses:go-gitea/gitea 2
GET commit-statuses:go-gitea/gitea
```

Of course this would require redis, but I think at some point it will be inevitable if we want to squeeze out the most performance with some caching tricks like this.

Disclaimer: I have no real knowledge of the underlying SQL structure. With proper database schema it's certainly possible to achieve similar performance with SQL as well.

File: routers/web/repo/repo.go
Comment (yardenshoham) [reply routers/web/repo/repo.go]: Now I batch the SQL queries

File: routers/web/repo/repo.go
Comment (yardenshoham) [reply routers/web/repo/repo.go]: I'm no longer making `O(n)` database calls, only `O(1)`. Is that good enough? (caching could be a future improvement)

Comment (silverwind) [reply routers/web/repo/repo.go]: O(1) is of course awesome 😆

Comment (silverwind) [reply routers/web/repo/repo.go]: BTW as for caching, it should also be possible to build a mini-redis into SQL by abusing a table of two string fields specifically for this purpose. So it may be possible to avoid a hard redis dependency.

**Comment by yardenshoham** [comment]
I made the code more performant, I think it's good enough to merge

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go]: Isn't that effectively
```suggestion
func GetLatestCommitStatusForPairs(ctx context.Context, repoIDsToLatestCommitSHAs map[int64]string, listOptions db.ListOptions) (map[int64][]*CommitStatus, error) {
```
?

File: routers/web/repo/repo.go
Comment (delvh) [inline routers/web/repo/repo.go]: If we already calculate the whole commit status, we might as well return the whole status instead of only its state…

File: routers/web/repo/repo.go
Comment (delvh) [inline routers/web/repo/repo.go]: This code confirms my suspicion above that it should be a map…

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go]:
```suggestion
		conds = append(conds, builder.Eq{"repo_id": repoIDs[i], "sha": shas[i]})
```

File: routers/web/repo/repo.go
Comment (lafriks) [inline routers/web/repo/repo.go]: It would be better if there was method to get default branch directly not get all branches and than loop

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go]:
```suggestion
	results := make([]Result, 0, len(repoIDs))
```

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go]: Why open a local type to the public?
(I mean the upper casing)

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go:262]: If I see this correctly, your code currently works because every subsequent  insert gets a higher ID, right?
I would've guessed that you want to sort by `created_unix DESC`.

Comment (delvh) [reply models/git/commit_status.go:262]: Furthermore, isn't your only goal the **latest** commit status of any repo?
So, why return a list of them for each repo then?
Is it because you can upload many different commit statuses for a single commit?
Even if yes, it's probably better to aggregate them, and only returning one combined status for each repo as that's what we want here.

File: models/git/commit_status.go
Comment (delvh) [inline models/git/commit_status.go]:
```suggestion
```

File: routers/web/repo/repo.go
Comment (yardenshoham) [reply routers/web/repo/repo.go]: Done

File: models/git/commit_status.go
Comment (yardenshoham) [reply models/git/commit_status.go]: fixed

File: routers/web/repo/repo.go
Comment (yardenshoham) [reply routers/web/repo/repo.go]: It would be tricky to add `CommitStatus` to the structs module because its defined in the git model. So because the entire status is not needed, I'd prefer to keep this as a future improvement

Comment (delvh) [reply routers/web/repo/repo.go]: Wouldn't this problem also be solved by moving your new `modules` structs into `services`?

Comment (yardenshoham) [reply routers/web/repo/repo.go]: Good thinking, done, thanks

File: models/git/commit_status.go
Comment (yardenshoham) [reply models/git/commit_status.go:262]: I was following the approach in https://github.com/go-gitea/gitea/blob/365bb77a5497d492836bf823cf780c823db27e8c/routers/web/repo/view.go#L836-L841

I will try to do what you suggest

Comment (yardenshoham) [reply models/git/commit_status.go:262]: Nevermind, I think we still need it because each repo can have more than one `context_hash` (like we have ~15 at the moment)
image

So can't simply grab the latest status

Comment (delvh) [reply models/git/commit_status.go:262]: See the last sentence which was exactly intended for this:
> Even if yes, it's probably better to aggregate them, and only returning one combined status for each repo as that's what we want here.

After all, there's already an API route that does the same, so it should be fairly easy.
In the end, the method should probably return a `map[int64]CommitStatus` instead of a `map[int64][]CommitStatus`.

Comment (yardenshoham) [reply models/git/commit_status.go:262]: Do you mean calling `git_model.CalcCommitStatus` inside `GetLatestCommitStatusForPairs`?

Comment (yardenshoham) [reply models/git/commit_status.go:262]: @delvh I'm not sure what you mean

Comment (lunny) [reply models/git/commit_status.go:262]: > If I see this correctly, your code currently works because every subsequent insert gets a higher ID, right? I would've guessed that you want to sort by `created_unix DESC`.

It's better to use id, created_unix maybe equal but id will not.

Comment (lunny) [reply models/git/commit_status.go:262]: ~~The problem here is the `branch`. The result should only for default branch of this repository. But maybe it hasn't been stored into `commit_status`.~~

~~Commit status of a pull request branch or a non-default branch may not be meaningful for the default branch.~~

~~I remember there is a feature request about the badge of CI, maybe it's related. Maybe this issue could depend on that one.~~

**Review by delvh — APPROVED** [review]
Let's hope the performance does not suffer too much for large instances.

**Comment by wxiaoguang** [comment]
The idea looks good to me. The only concern is the performance and whether the "status" could be used somewhere else when listing repositories.

For example, the new `GetBranch` seems inefficient, 30 repos means there are 30 git command calls.

There is a brief approach in my mind (I haven't thought about it carefully): what about adding a new column and save the check status into the repository table directly? Then when listing the repositories, a simple SQL query is enough.

**Comment by yardenshoham** [comment]
I'm not sure I'd know how to implement what you suggest. Aren't git calls generally cheap? This won't block the initial homepage render (it's another async call). There's also a limit so it wouldn't be too many repositories, I don't think it's too much of a performance hit.

If performance is the only concern, I suggest landing it as-is and doing further optimizations in future PRs.

**Comment by wxiaoguang** [comment]
> Aren't git calls generally cheap? 

Not cheap, it executes external git command. For example, there are similar issues:

* https://github.com/go-gitea/gitea/issues/20156#issuecomment-1168126474

> This won't block the initial homepage render (it's another async call).

It doesn't block page HTML loading, but it increases server load, and make the page rendering completion time increases (I haven't tested so I am not sure how much it would increase, but according to 20156, I think it is noticeable in most cases).

**Comment by yardenshoham** [comment]
Is this close to your vision? (still haven't implemented the repository table migration)
```diff
--- a/models/git/commit_status.go
+++ b/models/git/commit_status.go
@@ -361,6 +361,18 @@ func NewCommitStatus(ctx context.Context, opts NewCommitStatusOptions) error {
                return fmt.Errorf("insert CommitStatus[%s, %s]: %w", repoPath, opts.SHA, err)
        }
 
+       // get repo default branch and check if its commit sha is the same as the given sha
+       // if so, calculate the final commit status (git_model.CalcCommitStatus) and save it into the
+       // database table "repository"
+       branch, _ := git.GetBranch(ctx, opts.Repo.RepoPath(), opts.Repo.DefaultBranch)
+       commit, _ := branch.GetCommit()
+       if commit.ID.String() == opts.SHA {
+               commistStatuses, _, _ := GetLatestCommitStatus(ctx, opts.Repo.ID, opts.SHA, db.ListOptions{})
+               commitStatus := CalcCommitStatus(append(commistStatuses, opts.CommitStatus))
+               fmt.Println(commitStatus)
+               // TODO save commit status in database table "repository"
+       }
+
        return committer.Commit()
 }
```

**Comment by wxiaoguang** [comment]
> Is this close to your vision? (still haven't implemented the repository table migration)

Overall yes, but it seems not ideal to put these code in `models/git` package (I guess developers would expect `NewCommitStatus` only inserts a new record but doesn't do too much other logic, it's better not making a model package touch other model packages).

Ideally the logic seems more suitable to be put in `CreateCommitStatus` function.

----

However, did a quick code reading, I guess there seems more problems in the "commit status" related code (legacy problem)

One of the problems is: `NewCommitStatus` is called by 2 places, one is `actions.CreateCommitStatus` , another is `repo/files.CreateCommitStatus`, the logic is inconsistent, the worse thing is, the `MergeScheduledPullRequest` is not called everywhere and `files.CreateCommitStatus` only checks one status ..... the legacy code seems to have too many problems, I guess it should fix the legacy problems first.

Ideally after the fix, there should be one consistent `CreateCommitStatus` (in service package) to handle the commit status logic.

I am not sure whether my understanding is right, but without a clear design, there would be more maintainable problems sooner or later ..... so my opinion is if there is a chance, clean the legacy technical debts first?

And TBH, I do not want to be too strict, so if this PR looks good to other maintainers, I do not block.

**Comment by lunny** [comment]
30 times GetBranch looks really not good. I think this also depends on branch sync #22743 . Once branches has been stored into database, we can get default branch's last commit id from database. We can join the 3 tables to finish the query.

**Comment by yardenshoham** [comment]
@wxiaoguang if I add the suggested column like what I pasted above would you approve this PR? I do want to add this feature but a refactor is out of scope.

@lunny I understand and if #22743 is merged I would use it but it seems there is no progress there for months. I want this PR to be merged in 1.20

**Comment by wxiaoguang** [comment]
IMO,  add a new column (eg: default_branch_commit_status or some name you like) + update the column in service package when adding new commit status, it looks good to me and I can approve.

cc @lunny , do maintainers have concerns about this approach? I think this approach is efficient than joining more tables with 22743.

**Comment by lunny** [comment]
> IMO, add a new column (eg: default_branch_commit_status or some name you like) + update the column in service package when adding new commit status, it looks good to me and I can approve.
> 
> cc @lunny , do maintainers have concerns about this approach? I think this approach is efficient than joining more tables with 22743.

I think that's better than the current method.

**Comment by wxiaoguang** [comment]
Sorry for bothering, I realized that I made a mistake. My proposal (add a new column) doesn't work for edge cases: for example, if the default branch gets new commits but the commit status doesn't get updated, then users will see incorrect (inconsistent) commit status by that column. 

So, no idea from my side at the moment. Maybe it really needs to wait for some PRs like 22743 to sync the branches between database&git repo.

**Comment by yardenshoham** [comment]
Yes, I was in the middle of implementing this and you are right, I didn't think about that. So what do you think of landing this as-is? The default `RepoPagingNum` is 50 so it will never be more than 50 `GetBranch` (can we parallelize it?). It's probably not very noticeable

**Comment by wxiaoguang** [comment]
> So what do you think of landing this as-is?

Usually I only approve when I am >90% sure , since there is no conclusion about the performance problem, I haven't fully understand the problem ..... but that's only my opinion and I do not want to be too strict, so it depends the community's opinion (maybe call TOC?).

> The default RepoPagingNum is 50 so it will never be more than 50 GetBranch (can we parallelize it?).

I have no idea because I haven't tested. But according to 20156, around 100 calls cost 11 seconds, I guess 10 calls cost more than 1 seconds? For a large/busy instance (like Codeberg? I don't know, maybe 6543 knows), parallelizing might run out server resources more quickly.

**Comment by yardenshoham** [comment]
I have parallelized the git calls and cut them to 25% of what they were (commitID instead of branch->commitID->commit). I am migrating a few big git repositories to test the performance

**Comment by yardenshoham** [comment]
Serialized:
```
GetBranchCommitID on yardenshoham/kubernetes (size 2026697725) took 15.400108ms
GetBranchCommitID on yardenshoham/charts (size 1584540255) took 4.12132ms
GetBranchCommitID on yardenshoham/linux (size 5775060572) took 4.560219ms
GetBranchCommitID on yardenshoham/helm (size 85325852) took 7.65071ms
GetBranchCommitID on yardenshoham/gitea2 (size 748572595) took 4.449879ms
GetBranchCommitID on yardenshoham/gitea1 (size 746231760) took 3.0313ms
GetBranchCommitID on yardenshoham/act_runner (size 1221147) took 3.123049ms
GetBranchCommitID on yardenshoham/gitea (size 1150211519) took 3.692489ms
GetBranchCommitID on yardenshoham/test (size 33362) took 3.55677ms
GetBranchCommitID on yardenshoham/deno (size 288781961) took 3.551749ms
GetBranchCommitID on yardenshoham/actions-select (size 29320) took 2.97632ms
GetBranchCommitID on yardenshoham/.profile (size 29243) took 3.21083ms
GetBranchCommitID on yardenshoham/grogib (size 26190) took 3.319759ms
GetBranchCommitID on bigorg/grogib (size 26185) took 2.933869ms
```

I parallelized this (doesn't make much of a difference). In total, the request takes ~15ms (after the first time which takes ~70ms) with this PR, it was ~4ms before.

**Review by silverwind — APPROVED** [review]
Works. Perf is important, but given that this is a fetch call, it's not in the critical render path, so if this takes 1 second, it should still be acceptable.

**Comment by wxiaoguang** [comment]
@silverwind Please leave a large PR (especially when there are review comments) open for a few days before merge.

Although I do not block, but the new code (parallelizing) doesn't look good to me.

1. It shouldn't do so, parallelizing might run out server resources more quickly. Gitea shouldn't use a lot of go-routine in a web handler.
2. Data race! At least `repoIDsToLatestCommitSHAs[repo.ID] = commitID` is not right, there could be concurrent bugs.

I would suggest to revert this temporarily, and re-evaluate it by TOC

File: models/git/commit_status.go
Comment (wxiaoguang) [inline models/git/commit_status.go:261]: Outdated, there is `CalcCommitStatus` below, so it is right. 👍 

<details>

If I remember correctly, a commit could have multiple commit statuses?

For example:

* succeed status
* failed status
* succeed status


Then would this code consider a failed commit as succeeded?

</details>

**Comment by yardenshoham** [comment]
Why do you say "data race"? The execution order doesn't matter

**Comment by wxiaoguang** [comment]
Concurrent write to a map without lock is data race.

**Comment by silverwind** [comment]
> @silverwind Please leave a large PR (especially when there are review comments) open for a few days before merge.

Yeah sorry, I had a feeling this would spawn more comments when merging this 😉

**Comment by yardenshoham** [comment]
We'd have to replicate the templating logic in Vue

**Pull request merged**
